### PR TITLE
Better TF docstring types

### PR DESCRIPTION
**Pull request by Rocketknight1**
cc @sgugger - this is the PR based on the conversation we had on Slack! I just scanned our TF files and replaced some `Optional[]` and `Union[]` patterns with `|` instead. The doc builder now writes much cleaner docstrings, instead of the `tensorflow.python.framework.ops.Tensor` stuff it was writing before.

Some tests are failing because the test files also need `from __future__ import annotations` I think - will make sure everything passes before merging this.

### CODE REVIEW

**Comment by sgugger** [comment]
The tests seem to indicate we use those type annotations for something (auto-detecting the labels would be my guess).

**Comment by Rocketknight1** [comment]
We use them in one of the TF tests, in a way that causes issues with the `from __future__` import. I think there's an easy workaround for the affected test, though - working on it!

**Review by sgugger — APPROVED** [review]
LGTM but since it's such a big PR would love to have a second set of eyes on it. @amyeroberts could you quickly review the changes in:
- one of the modeling files,
- one of the test files
- in test_modeling_tf_common

I've reviewed all 139 files in the diff so no need to dive into all of them :-)

File: tests/test_modeling_tf_common.py [hunk 1]
```diff
@@ -1540,18 +1523,16 @@ def test_keras_fit(self):
...
             # We also remove "return_loss" as this is covered by the train_step when using fit()
             prepared_for_class = {
                 key: val
                 for key, val in prepared_for_class.items()
-                if key
-                not in ("head_mask", "decoder_head_mask", "cross_attn_head_mask", "decoder_input_ids", "return_loss")
+                if key not in ("head_mask", "decoder_head_mask", "cross_attn_head_mask", "return_loss")
             }
+            if "labels" in prepared_for_class and "decoder_input_ids" in prepared_for_class:
+                del prepared_for_class["decoder_input_ids"]
```
Comment (amyeroberts) [inline tests/test_modeling_tf_common.py:1535]: Just for my understanding - why do we now only remove `decoder_input_ids` if `labels` is present?

**Review by amyeroberts — APPROVED** [review]
Very nice - a lot cleaner 🤗 Thanks for adding!

File: tests/test_modeling_tf_common.py (hunk 1 above)
Comment (Rocketknight1) [reply tests/test_modeling_tf_common.py:1535]: Long story! That test previously had some weird return type logic to figure out which model classes actually returned losses and were trainable. Making this change broke that logic, but that logic was bad (and my fault) to begin with, so I refactored it. As a result, the test ran in some cases where it wasn't running before, which surfaced the bug that in some cases we were passing both `decoder_input_ids` and `labels` to conditional generation models that don't like that, and often have assertions that throw errors if you pass both. In general, for a basic test to check if `fit()` is working, never passing both of those at the same time seems to be the right choice for our models.

**Pull request merged**